<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"></ul></div>

Write a Python function called sqrt2 that calculates and prints to the screen the square root of 2 to 100 decimal places. Your code should not depend on any module from the standard library1 or otherwise. You should research the task first and include references and a description of your algorithm.

From Ian's video:
math.sqrt
https://docs.python.org/3/tutorial/floatingpoint.html
numpy
np.sqrt()
matplotlib.pyplot
https://tour.golang.org/flowcontrol/8 Newtons method
https://www.mathjax.org/ - built in to jupyter notebook (LaTeX math symbols)

https://catonmat.net/tools/generate-sqrt2-digits
Root 2 to 100 decimal places:
1.4142135623730950488016887242096980785696718753769480731766797379907324784621070388503875343276415727

The square root of 2 is also known as the one-half power of 2 https://en.wikipedia.org/wiki/Square_root_of_2

http://mathforum.org/library/drmath/view/65402.html
Proof:
(x^1/2) * (x^1/2) = x^(1/2 + 1/2) = x^1

However, when calculating the exponent below, Python returns the answer to just 16 decimal places, not the 100 required. 

In [1]:
2 ** .5

1.4142135623730951

This is due to the difficulty of computing floating point numbers https://docs.python.org/3.4/tutorial/floatingpoint.html, an issue that is not limited to the Python programming language. 

Figure out Newton Method of calculating square root.
Note example below different from the 2**.5 above. According to wiki the first few digits are https://catonmat.net/tools/generate-sqrt2-digits root 2 to 100 digits is 1.4142135623730950, which makes the above exponent calculation more accurate than the code below. 

In [4]:
# https://hackernoon.com/calculating-the-square-root-of-a-number-using-the-newton-raphson-method-a-how-to-guide-yr4e32zo
def mySqrt(x):

    r = x
    precision = 10 ** (-10)
    
    while abs(x - r * r) > precision:
        r = (r + x / r) / 2
        
    return r

mySqrt(2)

1.4142135623746899

Another method to overcome the floating point number limitations is to perform the calculation digit-by-digit
https://en.wikipedia.org/wiki/Methods_of_computing_square_roots#Digit-by-digit_calculation


The function below (taken from https://stackoverflow.com/a/5189881) works by first multiplying the number by 10^2(no. of decimal places required) and then getting the integer square root of that number


In [17]:
def sqroot(a, digits):
    a = a * (10**(2*digits))
    x_prev = 0
    x_next = 1 * (10**digits)
    while x_prev != x_next:
        x_prev = x_next
        x_next = (x_prev + (a // x_prev)) >> 1
    return x_next

x = sqroot(2, 100)
print(x)
len(str(x))


14142135623730950488016887242096980785696718753769480731766797379907324784621070388503875343276415727


101

Don't really understand above so trying to follow the algorithm section here: https://www.homeschoolmath.net/teaching/square-root-algorithm.php with help from https://www.youtube.com/watch?v=50oan6UYrX4&feature=emb_logo&ab_channel=SchoolMathsByThomasAnthikat
 
- This method uses number pairs to solve the square root so 2 would need to be written as 02.00000 (number of zeroes after decimal depends on how many decimal places you want in the answer. Each number pair represents a single digit in the solution
    - so 02.00 00 00 would give 4 digits of the solution (1.414)
    - this is why the code above has 2 * digits
    - it is multiplied by 10 to avoid working with floats
    


In [20]:
# find number whose square <=  2. This is the first number in the square root
x = int(2**.5)

# square that number and subtract from 2
y = 2-(x*x)
print(y)

# Double the 

1


Manually working through code 

def sqroot(2, 4):
    a = a * (10**(2*digits))                        a = 200000000   
    x_prev = 0                                      x_prev = 0              
    x_next = 1 * (10**digits)                       x_next = 10000          
    while x_prev != x_next:             
        x_prev = x_next                             x_prev = 10000   x_prev = 15000   x_prev = 14166   x_prev = 14142
        x_next = (x_prev + (a // x_prev)) >> 1      x_next = 15000   x_next = 14166   x_next = 14142   x_next = 14142
    return x_next


In [26]:
10000 + (200000000//10000)

30000

In [25]:
200000000//10000

20000

In [27]:
30000>>1

15000

In [31]:
(14142 + (200000000//14142)) >> 1

14142